esp-hal/src/dma/buffers.rs

```diff
@@ -1562,3 +1562,391 @@ impl DerefMut for DmaLoopBuf {
         self.buffer
     }
 }
+
+/// A custom DMA TX buffer that lets us unsafely fiddle with the descriptors.
+///
+/// Beware the dragons.
+///
+/// This type is only meant to fill out descriptors once, then send them off to the DMA. This
+/// isn't meant to hold onto the descriptors or otherwise provide a reuseable API to work with
+/// them. In fact this should just be a way to produce a `Preparation` but we have no way to do
+/// that in the current API.
+///
+/// This type likely never should be visible outside of esp-hal.
+pub(crate) struct UnsafeTxBuffer<'a> {
+    descriptors: DescriptorSet<'a>,
+    #[cfg(psram_dma)]
+    in_psram: bool,
+}
+
+impl<'a> UnsafeTxBuffer<'a> {
+    /// Creates a new DMA bufer.
+    ///
+    /// `block_size` is the requirement imposed by the peripheral that receives the data. It
+    /// ensures that the DMA will not try to copy a partial block, which would cause the RX DMA
+    /// to never complete.
+    ///
+    /// The function returns the DMA buffer, and the number of bytes that will be transferred.
+    ///
+    /// # Safety
+    ///
+    /// The caller must keep all its descriptors and the buffers they
+    /// point to valid while the buffer is being transferred.
+    #[cfg_attr(not(aes_dma), expect(unused))]
+    pub unsafe fn new(
+        descriptors: &'a mut [DmaDescriptor],
+        mut data: NonNull<[u8]>,
+        block_size: usize,
+    ) -> (Self, usize) {
+        let chunk_size = BurstConfig::DEFAULT
+            .max_chunk_size_for(unsafe { data.as_ref() }, TransferDirection::Out)
+            .min(4096 - block_size); // Whichever is stricter, data location or peripheral requirements
+
+        let data_len = data.len().min(chunk_size * descriptors.len());
+
+        cfg_if::cfg_if! {
+            if #[cfg(psram_dma)] {
+                let data_addr = data.addr().get();
+                let data_in_psram = crate::psram::psram_range().contains(&data_addr);
+
+                // Make sure input data is in RAM instead of cache
```

Comment:
Took me a second to understand this comment 😄
```suggestion
                // Make sure input data is in PSRAM instead of cache
```